configuration.md
## Configuration

Profiles can be edited by connecting a computer to the unit via USB and starting
a termnial program.  On my Linux laptop, I used:

    minicom -b 115200 -P /dev/ttyUSB0

There are two types of configuration: global and profile.

### Global Configuration

Some settings are always used, no matter what profile is selected.

#### Current calibration

The `ical` command is used to calibrate the current measurements.  It's
deault value is 133 milliOhms.  By measuring the current with a multimeter,
you can calibrate this value to match your specific hardware:

    ical 0.130

Would set the resistance to 130 milliOhms 

#### Voltage calibration

With a battery connected, measure the voltage at the vcal test point and enter
it with the `vcal` command.  For example:

    vcal 25.4

Note there is a related `vdrop` parameter but this parameter is  per-profile to
accommodate different breakout boards (e.g. whether they have protection diodes).

#### Voltage sag measurements

You can control how often to take voltage sag measurements.

First is `vsag_interval_seconds` which determines how often to take a baseline
measurement.  For example:

    vsag_interval_seconds 15

Next is `vsag_settle_ms` which determines how long to allow the battery to recover before taking a measurement.  For example:

    vsag_settle_ms 1000

#### Fan Settings

Use `fan_celsius <min_percent> <min_celsius> <max_celsius>` to control any connected fan(s)
At `min_celsius`, the fan(s) will run at `min_percent` and smoothly ramp to
100% speed at and beyond `max_celsius`. Example:

    fan_t 20 40 80 

Use `fan_watts <min_percent> <min_watts> <max_watts>` to control fans by power output.

The settings for `fan_t` and `fan_p` are active at the same time.  Whichever
leads to the higher fan speed at a given moment will be used.

Note that on may fans, using too low of a percent leads to no rotation.

#### Finish Display Time

Use `finish_display <mah_ratio>` to configure the number of seconds
that the finish stats are shown before the unit shuts down.  Example:

    finish_display 0.5

In the example above, if 1000 mAh were pulled from the battery then the display would be active for 1000 * 0.5 = 500 seconds.  If 100 mAh were pulled, then
the display would be active for 100 * 0.5 = 50 seconds.

### Responsiveness

The system looks to open up the FETs as much as possible while keeping under
the current limits of:

* Max voltage sag
* Max current
* Max power
* Max temperature

Thus the goal is to reach one of these maximums while staying under the maximum
for all others.  Which maximum is reached will vary by battery, fan settings, room
temperature, etc.

Coming up with and tuning an optimal PID algorithm that meets the above
requirements would be daunting due to the large number of free variables. For
example we would need to combine all four terms above to define an "error" as a
starting point, then would still need to determine the PID constants.

Fortunately, we do not need to system to converge in minimum time.  By relaxing
this requirement, we can get away with less complex calculations. 

The tuning algorithm is an annealing type algorithm where we use the terms
"velocity" and "acceleration" as a mental model.  As an analogy, imagine trying
to adjust the volume knob on a friend's stereo.  You would likely start with
some gross adjustment until you overshoot the mark, then hunt with increasingly
fine movements until you are satisfied.  If a "loud" commercial comes on,
you will need to adjust it again - again starting a bit fast and fine tuning.

With that in mind, here are the parameters of the algorithm:

* *max_velocity*  The maximum (and starting) velocity in percent / second
* *min_velocity* The minumum allow change velocity in percent / second.
* *deceleration_factor* the amount to decrease the velocity on an overshoot
  (going past the intended mark): from 0 to max_velocity
* *acceleration* the amount to increase the velocity (not reaching
  the intended mark) in percent / second. from 0.0 to max_velocity

Example:

  * fet_max_velocity = 30
  * fet_min_velocity = 0.05
  * fet_deceleration = 0.5
  * fet_acceleration = 0.1

  say the per sample time is 100ms

  The PWM will scale from 0 percent to 100% at 3% per sample, taking 33 samples
  or 3.3 seconds to get there.

  Say instead we hit a limit on the 10th sample.  PWN at this point would
  be 3 * 10 = 30%.  The velocity is reversed and decelerated, becoming
  to (-30 + 0.5) * 0.1 = -2.95% per sample

  On the next sample, PWM will change to 30 - 2.95 = 27.05%  Say that now
  everything is below threshold.  This means that we may have backed off
  too far.  Thus we reverse and decelerate again.  The new velocity becomes
  (29.5 - 0.5) * 0.1 = 2.9% per sample.

  One the next sample, PWM will change to 27.05 + 2.9 = 29.95.  Say that we
  are still too low.  In that case, the acceleration factor kicks in and
  the new velocity is not inverted, becoming (29 + 0.1) * 0.1 = 2.91% per sample.

  and so on.

### Profile Management

#### Create a profile

You can create a profile with the `new` command and a name.  For example:

    new

Profiles are named "New" as a starting point.  Alternately you can duplicate an
existing profile under a new name to use it's settings as a starting point:

    duplicate 0

would take whatever profile is at index zero and copy it under the next
available index.

#### Rename a profile

Change the name of a profile with the `name` command.  e.g.

    name 0 "LIPO 3.8 percell"

#### Delete a profile

Use the delete command with the profile index.  e.g.

    delete 1

#### Move a profile

Move a profile from one index to a different one via `move <source_idx> <dest_idx>`.  e.g.

    move 3 0

#### View profiles

Use the `list` command to view profile names and indexes:

    list

Use the `show` command to view settings for particular indexes or to dump all indexes:

    show
    show 0 2
    show 1 2 3

### Profile Configuration

#### Voltage drop configuration

Use `vdrop <profile_index> <voltage>` to set the voltage difference between the connected
battery and the vcal test point.  This will take into consideration various FETs and
diodes that exist between the two points. Example:

    vdrop 1 0.75

#### Cell count

Use `cell_count <profile_index> <count>` to adjust the number of cells.  If set at zero,
the number of cells will be automatically calculated at start by dividing the voltage
by the `per_cell_target_volts` (see below).  If the automatic calcualtion would lead to incorrect
results in your situation, you'll need to fix the value.

Note that you can also use a value of `1` if you don't want to think in terms of cells
but only the full voltage.  Examples:

    cell_count 1 0
    cell_count 1 1
    cell_count 1 6

#### Target voltage

Sets the per-cell target voltage.  When the target voltage is reached, the unit
will shut off.  For example:

    per_cell_target_volts 1 3.8

Would set the target voltage to 3.8 if `cell_count` is 1, 7.6 if `cell_count` is 2,
etc.

    per_cell_target_volts 1 0.0

Will run the drainer until it can no longer power itself, regardless of the cell count.

#### Damage voltage warning

If the target voltage is set below the `damage_voltage`, the user will be told that
they are about to destroy the connected batteries and will ask for confirmation.
The default settings is for LIPO batteries and is set to 3.3 volts.  Change it
with the `damage_voltage` command:

    damage_voltage 1 3.0

#### Max current

Use `max_amps <profile_index> <current>` to set the maximum allowed current.  Example

    max_amps 1 10.0

#### Max per-cell voltage sag

Use `per_cell_max_vsag <profile_index> <sag>` to determine the maximum amount of 
per-cell voltage sag that is allowed.  For example:

    per_cell_max_vsag 1 0.4

Would allow up to 0.4 * 4 = 1.6V of sag on a 4 cell pack.

#### Max temperature

Use `max_celsius <profile_index> <temp>` to change the maximum allowed heatsink termerature.  Example:

    max_celsius 1 80.0

#### Max Power

Use `max_watts <profile_indx> <watts>` to change the maximum allowed wattage (voltage * current).  Example:

    max_watts 1 150

### Testing

The console provides a `fake_mv` command.  If used, the voltage sense logic will provide that value instead
of the actual voltage reading.  It is useful for doing testing with no battery connected (since you can't
otherwise make it to the charging screen)  Using this setting with a battery connected is not recommended.


src/firmware/settings.c
#include "settings.h"
#include "hardware/flash.h"
#include "hardware/sync.h"

#include <string.h>
#include <stdio.h>


static inline size_t calc_sector_size(void) {
  const int num_sectors = 1 + (sizeof(struct Settings) / FLASH_SECTOR_SIZE);
  return num_sectors * FLASH_SECTOR_SIZE;
}

static inline size_t calc_page_size(void) {
  const int num_pages = 1 + (sizeof(struct Settings) / FLASH_PAGE_SIZE);
  return num_pages * FLASH_PAGE_SIZE;
}

static inline uint32_t flash_offset(void) {
  return (PICO_FLASH_SIZE_BYTES - calc_sector_size());
}

static inline uint8_t* flash_address(void) {
  return ((uint8_t*)(XIP_BASE + flash_offset()));
}

static uint16_t calc_checksum(const struct Settings* s) {
  const uint8_t* start = ((const uint8_t*)s) + sizeof(uint16_t);
  size_t len = sizeof(struct Settings) - sizeof(uint16_t);
  uint16_t sum = 0;
  for (size_t i=0; i<len; ++i) {
    sum += start[i] ^ i;
  }
  return sum;
}

// default resistance is 1000 low and 6800 + 2200 high
// at 30V we would have I = 30 / 10000 = 3 mA
// that would read as 0.003 * 1000 = 3V on the low side
// this would give an ADC reading of 4096.
// thus our default ratio is 30000 mv / 4096 ~= 7.3242
//#define DEFAULT_VCAL_RATIO 7.3242
#define DEFAULT_VCAL_RATIO 7.2464  // measured

static void global_settings_default(struct GlobalSettings* gs) {
  gs->ical_mohms = 133;
  gs->vcal_ratio = DEFAULT_VCAL_RATIO;

  gs->vsag.interval_seconds = 10;
  gs->vsag.settle_ms = 1000;

  gs->fan.min_percent = 20;
  gs->fan.min_celsius = 40;
  gs->fan.max_celsius = 70;
  gs->fan.min_watts = 5;
  gs->fan.max_watts = 70;

  gs->response.max_velocity = 10.0;
  gs->response.min_velocity = 0.05;
  gs->response.acceleration = 0.1;
  gs->response.deceleration = 0.5;

  gs->finish_display = 1.0;
}

static void init_profile(struct ProfileSettings* ps) {
  strcpy(ps->name, "Default");
  ps->drop_mv = 0;
  ps->cell_count = 0;

  ps->cell.target_mv = 3800;
  ps->cell.damage_mv = 3300;
  ps->cell.max_vsag_mv = 300;

  ps->max_ma = 5000;
  ps->max_celsius = 80;
  ps->max_watts = 100;
}

void settings_try_add_profile(struct Settings* settings) {
  if (settings->profile_count >= MAX_PROFILE_COUNT) {
    printf("Can not add a profile: max allowed is MAX_PROFILE_COUNT\n");
    return;
  }

  struct ProfileSettings* ps = settings->profile + settings->profile_count;
  init_profile(ps);
  printf("Added profile: %d %s (not saved)\n", settings->profile_count, ps->name);
  ++settings->profile_count;
}

void settings_try_duplicate_profile(struct Settings* settings, uint8_t source_idx) {
  if (source_idx >= settings->profile_count) {
    printf("Invalid source profile index\n");
    return;
  }
  if (settings->profile_count >= MAX_PROFILE_COUNT) {
    printf("Can not add a profile: max allowed is MAX_PROFILE_COUNT\n");
    return;
  }

  const struct ProfileSettings* source_ps = settings->profile + source_idx;
  struct ProfileSettings* dest_ps = settings->profile + settings->profile_count;
  memcpy(dest_ps, source_ps, sizeof(struct ProfileSettings));
  printf("Duplicated profile: %d %s -> %d (not saved)\n", source_idx, dest_ps->name, settings->profile_count);
  ++settings->profile_count;
}

void settings_try_delete_profile(struct Settings* settings, uint8_t idx) {
  if (idx >= settings->profile_count) {
    printf("Invalid profile index\n");
    return;
  }
  if (settings->profile_count <= 1) {
    printf("At least one profile must be defined\n");
    return;
  }

  // say there are 5 profile and we want to remove #1
  // Thus we had 0, 1, 2, 3, 4 and want to have 0, 2, 3, 4
  //
  // copy length is 5 - 1 - 1 = 3
  const size_t copy_length = sizeof(struct ProfileSettings) * (settings->profile_count - idx - 1);
  if (copy_length > 1) {
    memmove(settings->profile + idx, settings->profile + idx + 1, copy_length);
  }
  --settings->profile_count;

  // Clear out the now-unused slot just for tidiness
  memset(settings->profile + settings->profile_count, 0, sizeof(struct ProfileSettings));
  printf("Removed profile %d (not saved)\n", idx);
}

void settings_try_move_profile(struct Settings* settings, uint8_t src_idx, uint8_t dest_idx) {
  if (src_idx >= settings->profile_count) {
    printf("Invalid source index\n");
    return;
  }
  if (dest_idx >= settings->profile_count) {
    printf("Invalid destination index\n");
    return;
  }
  if (src_idx == dest_idx) {
    printf("source == destination index\n");
    return;
  }

  // for out example, we will say we have 5 slots and want to move slot 1 to slot 3
  // e.g. from [0 1 2 3 4] to [0 2 3 1 4]

  // first, make a backup of the source profile
  struct ProfileSettings src_ps;
  memcpy(&src_ps, settings->profile + src_idx, sizeof(struct ProfileSettings));

  // next delete the source slot
  // Here we go from [0 1 2 3 4] to [0 2 3 4] with 1 backed up.
  size_t copy_length = sizeof(struct ProfileSettings) * (settings->profile_count - src_idx - 1);
  if (copy_length > 1) {
    memmove(settings->profile + src_idx, settings->profile + src_idx + 1, copy_length);
  }

  // now we want to make an opening at dest_idx
  // Here we go from [0 2 3 4] to [0 2 3 _ 4]
  copy_length = sizeof(struct ProfileSettings) * (settings->profile_count - dest_idx - 1);
  if (copy_length > 1) {
    memmove(settings->profile + dest_idx + 1, settings->profile + dest_idx, copy_length);
  }

  // finally, copy the backed-up profile into the empty slot
  memcpy(settings->profile + dest_idx, &src_ps, sizeof(struct ProfileSettings));
  printf("Moved profile %d to index %d (not saved)\n", src_idx, dest_idx);
}


void settings_default(struct Settings* settings) {
  memset(settings, 0, sizeof(struct Settings));
  settings->eyecatcher[0] = 'B';
  settings->eyecatcher[0] = 'A';
  settings->eyecatcher[0] = 'T';
  settings->eyecatcher[0] = 'D';

  settings->profile_count = 0;
  global_settings_default(&(settings->global));
  settings_try_add_profile(settings);
  printf("Settings set to default values\n");
}

void settings_load(struct Settings* settings) {
  memcpy(settings, flash_address(), sizeof(struct Settings));
  const uint16_t checksum = calc_checksum(settings);
  if (checksum == settings->checksum) {
    printf("Settings loaded from flash\n");
  } else {
    settings_default(settings);
  }
}

// Do not call save() excessively as it could wear out the
// flash memory.
void settings_save(struct Settings* settings) {
  settings->checksum = calc_checksum(settings);
  uint32_t ints = save_and_disable_interrupts();
  flash_range_erase(flash_offset(), calc_sector_size());
  flash_range_program(flash_offset(), (const uint8_t*)settings, calc_page_size());
  restore_interrupts(ints);
  printf("Settings saved to flash.\n");
}

void settings_calc_voltage_and_cell_count(
    const struct Settings* settings,
    int profile_index,
    uint16_t current_mv,
    uint8_t* cell_count,
    uint16_t* target_mv) {
  *cell_count = 0;
  *target_mv = 0;
  if (profile_index >= settings->profile_count) {
    return;
  }
  const struct ProfileSettings* ps = settings->profile + profile_index;
  if (ps->cell.target_mv == 0) {
    return;
  }
  uint32_t cells = ps->cell_count > 0 ? ps->cell_count : current_mv / (ps->cell.target_mv - 150);
  if (cells > 6) {
      // Target_mv must be too low for the calculation to work
    return;
  }
  *cell_count = (uint8_t)cells;
  *target_mv = (uint16_t)(cells * ps->cell.target_mv);
}

